### PR TITLE
Add `step: query` to e2e tests for SELECT queries

### PR DESCRIPTION
**Pull request by pkong-ds**
## What's in this PR?
Add a new step to run a database query and assert the result

## Why?
1. reauth need fetch user_id from db to `generateIDToken`
2. userimport needs to verify user attributes are populated correctly (uppercase/lowercase) in db

## Usage example
Please see https://github.com/authgear/authgear-server/pull/4562/commits/28f884a9f658ca279b21bd056c98d6f06908c618 at PR #4562

ref DEV-1253

### CODE REVIEW

File: e2e/cmd/e2e/pkg/sql.go [hunk 1]
```diff
@@ -12,3 +13,75 @@ func ParseSQLTemplate(name string, rawTemplateText string) (*template.Template,
...
+		return
+	}
+
+	nCol := len(columnTypes)
+
+	for rows.Next() {
+		scanArgs := make([]interface{}, nCol)
+		for i, v := range columnTypes {
+			switch v.DatabaseTypeName() {
+			case "VARCHAR", "TEXT", "UUID", "TIMESTAMP":
```
Comment (louischan-oursky) [inline e2e/cmd/e2e/pkg/sql.go:31]: TIMESTAMP should be `*time.Time`

File: e2e/pkg/testrunner/testcase.go [hunk 2]
```diff
@@ -219,6 +219,37 @@ func (tc *TestCase) executeStep(
 			Error: nil,
 		}
 
+	case StepActionQuery:
+		jsonArrString, err := cmd.QuerySQLSelectFile(step.Query)
+		if err != nil {
+			t.Errorf("failed to execute SQL Select query: %v", err)
+			return
+		}
+		fmt.Printf("pkong#jsonArrString: %+v\n", jsonArrString)
```
Comment (louischan-oursky) [inline e2e/pkg/testrunner/testcase.go:228]: Remove this line.

File: e2e/pkg/testrunner/testcase.go [hunk 3]
```diff
@@ -219,6 +219,37 @@ func (tc *TestCase) executeStep(
...
+		if err != nil {
+			t.Errorf("failed to unmarshal json rows: %v", err)
+			return
+		}
+		rowsResult["rows"] = rows
+		result = &StepResult{
+			Result: rowsResult,
+			Error:  nil,
+		}
+		fmt.Printf("pkong#result: %+v\n", result)
```
Comment (louischan-oursky) [inline e2e/pkg/testrunner/testcase.go:242]: Remove this line.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

File: e2e/cmd/e2e/pkg/sql.go [hunk 4]
```diff
@@ -77,6 +79,11 @@ func ParseRows(rows *sql.Rows) (outputRows []interface{}, err error) {
 				continue
 			}
 
+			if z, ok := (scanArgs[i]).(*sql.NullTime); ok {
+				row[v.Name()] = z.Time
+				continue
+			}
+
```
Comment (louischan-oursky) [inline e2e/cmd/e2e/pkg/sql.go:86]: If we write this, then we cannot distinguish between null and zero value. We should output a JSON null if the value is a SQL NULL.

File: e2e/README.md [hunk 5]
```diff
@@ -186,6 +187,36 @@ The `oauth_redirect` action is used to redirect to an OAuth provider. The result
...
+# get_my_user.sql
+SELECT id
+FROM _auth_user 
+WHERE app_id = '{{ .AppID }}'
+AND standard_attributes ->> 'preferred_username' = 'my_username';
+```
+
+```yaml
+- action: query
+  query: get_my_user.sql
```
Comment (louischan-oursky) [inline e2e/README.md:204]: Change to a string. We still support template.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by tung2744** [comment]
This seems something similar to https://linear.app/authgear/issue/DEV-1253/[e2e-testing]-support-action-type-assertion
so I also linked the pr to the issue.

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

File: e2e/cmd/e2e/pkg/sql.go [hunk 6]
```diff
@@ -55,32 +55,56 @@ func ParseRows(rows *sql.Rows) (outputRows []interface{}, err error) {
...
 				continue
 			}
 
 			if z, ok := (scanArgs[i]).(*sql.NullTime); ok {
-				row[v.Name()] = z.Time
+				if z.Valid { // not null
+					row[v.Name()] = z.Time
+				} else { // null
+					row[v.Name()] = nil
+				}
```
Comment (louischan-oursky) [inline e2e/cmd/e2e/pkg/sql.go:107]: In line 111, we should panic. The rationale is that we should have handled all data types we have recognized.

File: e2e/pkg/testrunner/cmd.go [hunk 7]
```diff
@@ -51,6 +51,16 @@ func (e *End2EndCmd) ExecuteCustomSQL(sqlPath string) error {
 	return nil
 }
 
+func (e *End2EndCmd) QuerySQLSelectFile(rawSQL string) (jsonArrString string, err error) {
```
Comment (louischan-oursky) [inline e2e/pkg/testrunner/cmd.go:54]: Rename QuerySQLSelectFile to QuerySQLSelectRaw

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**